File: cpp/bytecode-vm/src/lexer.cpp
#include "lexer.hpp"

#include <cctype>
#include <unordered_map>

namespace {
const std::unordered_map<std::string, TokenType> kKeywords = {
    {"var", TokenType::Var}, {"if", TokenType::If}, {"else", TokenType::Else},
    {"while", TokenType::While}, {"fun", TokenType::Fun}, {"return", TokenType::Return},
    {"print", TokenType::Print}, {"true", TokenType::True}, {"false", TokenType::False},
    {"nil", TokenType::Nil}, {"and", TokenType::And}, {"or", TokenType::Or},
};
}

Lexer::Lexer(std::string source) : source_(std::move(source)) {}

bool Lexer::isAtEnd() const { return current_ >= source_.size(); }

char Lexer::advance() { return source_[current_++]; }

char Lexer::peek() const { return isAtEnd() ? '\0' : source_[current_]; }

char Lexer::peekNext() const {
    if (current_ + 1 >= source_.size()) return '\0';
    return source_[current_ + 1];
}

bool Lexer::match(char expected) {
    if (isAtEnd() || source_[current_] != expected) return false;
    current_++;
    return true;
}

Token Lexer::makeToken(TokenType type) const {
    return Token{type, source_.substr(start_, current_ - start_), line_};
}

Token Lexer::errorToken(const std::string& message) const {
    return Token{TokenType::Error, message, line_};
}

void Lexer::skipWhitespace() {
    for (;;) {
        char c = peek();
        switch (c) {
            case ' ':
            case '\r':
            case '\t':
                advance();
                break;
            case '\n':
                line_++;
                advance();
                break;
            case '/':
                if (peekNext() == '/') {
                    while (peek() != '\n' && !isAtEnd()) advance();
                } else {
                    return;
                }
                break;
            default:
                return;
        }
    }
}

Token Lexer::scanString() {
    while (peek() != '"' && !isAtEnd()) {
        if (peek() == '\n') line_++;
        advance();
    }
    if (isAtEnd()) return errorToken("Unterminated string.");
    advance();
    Token t = makeToken(TokenType::String);
    t.lexeme = source_.substr(start_ + 1, (current_ - start_) - 2);
    return t;
}

Token Lexer::scanNumber() {
    while (std::isdigit(static_cast<unsigned char>(peek()))) advance();
    if (peek() == '.' && std::isdigit(static_cast<unsigned char>(peekNext()))) {
        advance();
        while (std::isdigit(static_cast<unsigned char>(peek()))) advance();
    }
    return makeToken(TokenType::Number);
}

Token Lexer::scanIdentifier() {
    while (std::isalnum(static_cast<unsigned char>(peek())) || peek() == '_') advance();
    std::string text = source_.substr(start_, current_ - start_);
    auto it = kKeywords.find(text);
    if (it != kKeywords.end()) return makeToken(it->second);
    return makeToken(TokenType::Identifier);
}

Token Lexer::scanToken() {
    skipWhitespace();
    start_ = current_;
    if (isAtEnd()) return makeToken(TokenType::Eof);

    char c = advance();
    if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') return scanIdentifier();
    if (std::isdigit(static_cast<unsigned char>(c))) return scanNumber();

    switch (c) {
        case '(': return makeToken(TokenType::LParen);
        case ')': return makeToken(TokenType::RParen);
        case '{': return makeToken(TokenType::LBrace);
        case '}': return makeToken(TokenType::RBrace);
        case ',': return makeToken(TokenType::Comma);
        case ';': return makeToken(TokenType::Semicolon);
        case '+': return makeToken(TokenType::Plus);
        case '-': return makeToken(TokenType::Minus);
        case '*': return makeToken(TokenType::Star);
        case '/': return makeToken(TokenType::Slash);
        case '!': return makeToken(match('=') ? TokenType::BangEqual : TokenType::Bang);
        case '=': return makeToken(match('=') ? TokenType::EqualEqual : TokenType::Equal);
        case '<': return makeToken(match('=') ? TokenType::LessEqual : TokenType::Less);
        case '>': return makeToken(match('=') ? TokenType::GreaterEqual : TokenType::Greater);
        case '"': return scanString();
    }
    return errorToken("Unexpected character.");
}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;
    for (;;) {
        Token t = scanToken();
        tokens.push_back(t);
        if (t.type == TokenType::Eof || t.type == TokenType::Error) break;
    }
    return tokens;
}

File: python/typed-lang/lexer.py
from dataclasses import dataclass


KEYWORDS = {
    "var", "if", "else", "while", "def", "return",
    "print", "true", "false", "and", "or", "not",
    "int", "float", "bool", "string", "void",
}


@dataclass
class Token:
    type: str
    value: str
    line: int


class LexError(Exception):
    pass


SIMPLE_TOKENS = {
    "(": "LPAREN", ")": "RPAREN", "{": "LBRACE", "}": "RBRACE",
    ",": "COMMA", ";": "SEMI", ":": "COLON",
    "+": "PLUS", "-": "MINUS", "*": "STAR", "/": "SLASH",
}


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1

    def peek(self, offset: int = 0) -> str:
        idx = self.pos + offset
        return self.source[idx] if idx < len(self.source) else "\0"

    def advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
        return ch

    def tokenize(self) -> list[Token]:
        tokens: list[Token] = []
        while self.pos < len(self.source):
            ch = self.peek()

            if ch in " \r\t\n":
                self.advance()
                continue

            if ch == "/" and self.peek(1) == "/":
                while self.peek() != "\n" and self.pos < len(self.source):
                    self.advance()
                continue

            if ch.isdigit():
                tokens.append(self._number())
                continue

            if ch.isalpha() or ch == "_":
                tokens.append(self._identifier())
                continue

            if ch == '"':
                tokens.append(self._string())
                continue

            line = self.line
            if ch == "=" and self.peek(1) == "=":
                self.advance(); self.advance()
                tokens.append(Token("EQEQ", "==", line))
                continue
            if ch == "!" and self.peek(1) == "=":
                self.advance(); self.advance()
                tokens.append(Token("NEQ", "!=", line))
                continue
            if ch == "<" and self.peek(1) == "=":
                self.advance(); self.advance()
                tokens.append(Token("LE", "<=", line))
                continue
            if ch == ">" and self.peek(1) == "=":
                self.advance(); self.advance()
                tokens.append(Token("GE", ">=", line))
                continue
            if ch == "-" and self.peek(1) == ">":
                self.advance(); self.advance()
                tokens.append(Token("ARROW", "->", line))
                continue
            if ch == "=":
                self.advance()
                tokens.append(Token("EQ", "=", line))
                continue
            if ch == "<":
                self.advance()
                tokens.append(Token("LT", "<", line))
                continue
            if ch == ">":
                self.advance()
                tokens.append(Token("GT", ">", line))
                continue

            if ch in SIMPLE_TOKENS:
                self.advance()
                tokens.append(Token(SIMPLE_TOKENS[ch], ch, line))
                continue

            raise LexError(f"Unexpected character {ch!r} at line {line}")

        tokens.append(Token("EOF", "", self.line))
        return tokens

    def _number(self) -> Token:
        line = self.line
        start = self.pos
        is_float = False
        while self.peek().isdigit():
            self.advance()
        if self.peek() == "." and self.peek(1).isdigit():
            is_float = True
            self.advance()
            while self.peek().isdigit():
                self.advance()
        text = self.source[start:self.pos]
        return Token("FLOAT" if is_float else "INT", text, line)

    def _identifier(self) -> Token:
        line = self.line
        start = self.pos
        while self.peek().isalnum() or self.peek() == "_":
            self.advance()
        text = self.source[start:self.pos]
        if text in KEYWORDS:
            return Token(text.upper(), text, line)
        return Token("IDENT", text, line)

    def _string(self) -> Token:
        line = self.line
        self.advance()
        start = self.pos
        while self.peek() != '"':
            if self.pos >= len(self.source):
                raise LexError(f"Unterminated string at line {line}")
            self.advance()
        text = self.source[start:self.pos]
        self.advance()
        return Token("STRING", text, line)
